File: sdk/r/R/cloud_service.R
#' Cloud Service: VM shutdown, reset, hostname, IP pools, networks, interfaces
#' @param http HttpClient object
#' @return Resource object
#' @export
CloudServiceResource <- function(http) {
  env <- new.env(parent = emptyenv())
  env$http <- http
  class(env) <- "CloudServiceResource"
  env
}

#' POST /api/service/{id}/vms/{vmid}/shutdown - Perform graceful shutdown
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
shutdownVm <- function(self, ...) {
  self$http$post("/api/service/{id}/vms/{vmid}/shutdown", list(...))
}

#' POST /api/service/{id}/vms/{vmid}/reset - Reset virtual server power
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
resetVm <- function(self, ...) {
  self$http$post("/api/service/{id}/vms/{vmid}/reset", list(...))
}

#' POST /api/service/{id}/vms/{vmid}/hostname - Update VM hostname
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
changeHostname <- function(self, ...) {
  self$http$post("/api/service/{id}/vms/{vmid}/hostname", list(...))
}

#' GET /api/service/{id}/vms/{vmid}/ippool - List IP pools
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
listIpPools <- function(self, ...) {
  self$http$get("/api/service/{id}/vms/{vmid}/ippool", list(...))
}

#' POST /api/service/{id}/vms/{vmid}/ippool/{pool} - Allocate new IP
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
allocateIp <- function(self, ...) {
  self$http$post("/api/service/{id}/vms/{vmid}/ippool/{pool}", list(...))
}

#' GET /api/service/{id}/networks - List available networks
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
listNetworks <- function(self, ...) {
  self$http$get("/api/service/{id}/networks", list(...))
}

#' POST /api/service/{id}/vms/{vmid}/interfaces - Add network interface
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
addInterface <- function(self, ...) {
  self$http$post("/api/service/{id}/vms/{vmid}/interfaces", list(...))
}

#' GET /api/service/{id}/vms/{vmid}/interfaces/{iface} - Get interface details
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
getInterface <- function(self, ...) {
  self$http$get("/api/service/{id}/vms/{vmid}/interfaces/{iface}", list(...))
}

#' PUT /api/service/{id}/vms/{vmid}/interfaces/{iface} - Update interface
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
updateInterface <- function(self, ...) {
  self$http$put("/api/service/{id}/vms/{vmid}/interfaces/{iface}", list(...))
}

#' DELETE /api/service/{id}/vms/{vmid}/interfaces/{iface} - Remove interface
#' @param self Resource object
#' @param ... Additional parameters
#' @return API response
removeInterface <- function(self, ...) {
  self$http$delete("/api/service/{id}/vms/{vmid}/interfaces/{iface}", list(...))
}

File: sdk/c/src/cloud_service.c
#include "hitechcloud/cloud_service.h"
#include <stdlib.h>


/* POST /api/service/{id}/vms/{vmid}/shutdown - Perform graceful shutdown */
char *hc_cloud_service_shutdown_vm(hc_http_client_t *client) {
    return hc_post(client, "/api/service/{id}/vms/{vmid}/shutdown", NULL);
}

/* POST /api/service/{id}/vms/{vmid}/reset - Reset virtual server power */
char *hc_cloud_service_reset_vm(hc_http_client_t *client) {
    return hc_post(client, "/api/service/{id}/vms/{vmid}/reset", NULL);
}

/* POST /api/service/{id}/vms/{vmid}/hostname - Update VM hostname */
char *hc_cloud_service_change_hostname(hc_http_client_t *client) {
    return hc_post(client, "/api/service/{id}/vms/{vmid}/hostname", NULL);
}

/* GET /api/service/{id}/vms/{vmid}/ippool - List IP pools */
char *hc_cloud_service_list_ip_pools(hc_http_client_t *client) {
    return hc_get(client, "/api/service/{id}/vms/{vmid}/ippool", NULL);
}

/* POST /api/service/{id}/vms/{vmid}/ippool/{pool} - Allocate new IP */
char *hc_cloud_service_allocate_ip(hc_http_client_t *client) {
    return hc_post(client, "/api/service/{id}/vms/{vmid}/ippool/{pool}", NULL);
}

/* GET /api/service/{id}/networks - List available networks */
char *hc_cloud_service_list_networks(hc_http_client_t *client) {
    return hc_get(client, "/api/service/{id}/networks", NULL);
}

/* POST /api/service/{id}/vms/{vmid}/interfaces - Add network interface */
char *hc_cloud_service_add_interface(hc_http_client_t *client) {
    return hc_post(client, "/api/service/{id}/vms/{vmid}/interfaces", NULL);
}

/* GET /api/service/{id}/vms/{vmid}/interfaces/{iface} - Get interface details */
char *hc_cloud_service_get_interface(hc_http_client_t *client) {
    return hc_get(client, "/api/service/{id}/vms/{vmid}/interfaces/{iface}", NULL);
}

/* PUT /api/service/{id}/vms/{vmid}/interfaces/{iface} - Update interface */
char *hc_cloud_service_update_interface(hc_http_client_t *client) {
    return hc_put(client, "/api/service/{id}/vms/{vmid}/interfaces/{iface}", NULL);
}

/* DELETE /api/service/{id}/vms/{vmid}/interfaces/{iface} - Remove interface */
char *hc_cloud_service_remove_interface(hc_http_client_t *client) {
    return hc_delete(client, "/api/service/{id}/vms/{vmid}/interfaces/{iface}", NULL);
}

